### ISSUE TITLE
Invalid grant: Password grant is disabled for insecure clients

### ISSUE THREAD

**Issue by vdbrink** [elisaado/somtoday-api-docs#24]
Sinds vanavond werkt het opvragen van de token niet meer met `grant_type=password`. De error die ik krijg is `Invalid grant: Password grant is disabled for insecure clients.`

Heeft er iemand al een alternatieve manier voor de authenticatie?

**Comment by elisaado** [comment]
oh God help

**Comment by 770grappenmaker** [comment]
De somtoday app gebruikt al langere tijd de "inloggen met somtoday" knop. Misschien is soortgelijk gebruik nu verplicht? Kleine check: heb je wel de client id meegegeven?

**Comment by vdbrink** [comment]
Client id stuur ik mee. Tot gisteravond werkte het ook nog prima. Nu waarschijnlijk overgestapt op een ander grant type op basis van een authenticatie code ipv username /password

**Comment by elisaado** [comment]
lmao somtoday heeft zichzelf genuked.

Maar ik denk dat ze inderdaad single sign on als enige inlogmethode willen pushen.

**Comment by Fish-o** [comment]
kan dit reproduceren, waarschijnlijk inderdaad iets met sso. Zal vanavond kijken hoe we dit kunnen aanpakken.

**Comment by peterj1977** [comment]
Wie kan mij helpen om gegevens van Somtoday te implementeren in Home Assistant, zodat ik de cijfers, huiswerk en rooster kan showen.

**Comment by 770grappenmaker** [comment]
Ik kan nog wel "inloggen met somtoday"? Het zal wel verschillen per school. Vooralsnog lijkt het me dan dat somtoday een foutje heeft gemaakt inderdaad.

**Comment by 25huizengek1** [comment]
Ik heb hier een fix voor gevonden. Als je in de sourcecode van de Somtoday Leerling app duikt, kun je erachter komen hoe dit werkt. Voor de client id is er namelijk nu een redirect uri bekend: `somtodayleerling://oauth/callback`, als je nu naar `https://inloggen.somtoday.nl/oauth2/authorize?client_id=D50E0C06-32D1-4B41-A137-A9A850C892C2&response_type=code&scope=openid&redirect_uri=somtodayleerling%3A%2F%2Foauth%2Fcallback&tenant_uuid=(schoolid)` gaat, zul je zien dat de browser daarna naar `somtodayleerling://oauth:443/callback?code=(code)&iss=https://somtoday.nl` zal gaan. Als je dat afwacht, kun je daarna een request sturen naar `https://somtoday.nl/oauth2/token`, maar dit keer met als `grant_type` `authorization_code`, `session` moet `no_session` zijn, er moet een `tenant_uuid` zijn, en verder blijft alles hetzelfde, behalve natuurlijk gebruikersnaam en wachtwoord.

**Comment by Fish-o** [comment]
Dag @25huizengek1 !
Een tijdje geleden heb ik een fork gemaakt met en een mogelijke fix gedocumenteerd.
Het lijkt overeen te komen met wat jij hier meld. Kan je kijken of wat er staat correct is en ook voor jou werkt? Dan kan ie gemerged worden!

https://github.com/Fish-o/somtoday-api-docs

**Comment by 25huizengek1** [comment]
Dag @Fish-o 
Wat ik in de gelinkte documentatie terugzie is een "browserloze" manier om in te loggen. Hier had ik niet over nagedacht, aangezien ik zelf maar accepteerde dat er een browser voor nodig was, want ik had toch al een UI in mijn code zitten. Mijn manier is vrijwel gelijk (op de state etc na), maar het verzenden van het formulier heb je ook gelijk gedocumenteerd, waardoor er dus geen browser vereist is. Dit lijkt me dus goede documentatie, behalve dat ik het niet kon reproduceren. Dit is waarschijnlijk omdat het portal nu een andere inlogflow heeft, met eerst gebruikersnaam, submit, en daarna pas wachtwoord.

Ik ging dus in de HTML kijken, en al vrij gauw kwam ik erachter dat op deze manier eerst een gebruikersnaam en daarna een wachtwoord wordt verwacht door het portaal. Ik haalde het field voor het wachtwoord weg en voilà, het werkte. Als je daarna een request stuurt als volgt: `POST /login?1-1.-passwordForm&auth=(auth)` met voor de rest gewoon hetzelfde als bij username, maar dan zonder de username weer opnieuw mee te geven, dan krijg je de code terug van de server. Het is misschien wel goed om mee te delen dat de cookies die bij het username request worden ontvangen, die ook weer terug naar de server moeten bij het wachtwoord. Wat ook verwarrend kan zijn is het volgende: als gebruikersnaam of wachtwoord niet overeenkomt, krijg je 200 OK, als het klopt juist 302 Found. Bij het verifiëren van de code is het weer andersom: bij success 200, anders 400.

Als deze dingen worden toegevoegd, kan er wat mij betreft een pull request komen.

**Comment by Fish-o** [comment]
@25huizengek1 
Ben een beetje druk geweest, maar heb nu weer wat tijd!
Voor mij werkt deze proof of concept, die ik destijdig heb gemaakt nog steeds:
https://gist.github.com/Fish-o/83f06bdd9f412c41424698f9352fed47
Ik stuur hiering de gebruikersnaam en wachtwoord tegelijkertijd.

De auth flow kan per school verschillen, en dus mischien werkt het voor jou dus ook wel anders.
Kan het mischien zijn dat je school gebruik maakt van Microsoft SSO?

De reden dat de eerste request een 302 is, is omdat het een redirect is naar de inlog pagina. Bij de laatste stap is het ook weer een redirect, maar deze keer naar somtodayleerling://oauth/callback (de app op je telefoon).

**Comment by 25huizengek1** [comment]
Nee, mijn school maakt daar geen gebruik van. Gewoon gebruikersnaam en wachtwoord invoeren in som zelf. Alleen moet je eerst je gebruikersnaam invullen, op een knop klikken, en daarna wachtwoord.

**Comment by Fish-o** [comment]
Dus die POC werkt niet voor jou?

**Comment by 25huizengek1** [comment]
Hij stuurt me door naar `https://inloggen.somtoday.nl/login?1`, volgens het loggen van de response`, en laat dat nou een formulier zijn voor het invullen van het wachtwoord.... (voorbeeld via browser)
![image](https://user-images.githubusercontent.com/50515369/169062025-e6bd78fa-e979-4e4a-a96a-b18c33c13016.png)

**Comment by 25huizengek1** [comment]
@Fish-o Enige updates over deze issue? Probeer anders zelf nog eens je eigen POC uit te voeren, want het zou kunnen zijn dat het een tijdje geleden wel prima werkte, maar nu niet meer.

**Comment by Fish-o** [comment]
Nope, werkt nog steeds voor mij. 
Vult ie het nog steeds automatisch in op een andere browser, of als je je cookies en dergelijken verwijdert?
De docs volgen trouwens de inlog flow van de (android) app, als je via daar inlogt vult ie het dan ook automatisch in?

**Comment by 25huizengek1** [comment]
Ik heb hier nog even over na zitten denken en ben met het geniale idee gekomen om mijn POC en [deze](https://gist.github.com/Fish-o/83f06bdd9f412c41424698f9352fed47) samen te voegen in een apart programma om te testen wat er wel/niet werkt. https://github.com/25huizengek1/somtoday-poc

Mijn output komt erop neer dat mijn code wel werkt, maar als het werkelijk per school verschilt zou er bij jou wat anders uit moeten komen.

**Comment by 25huizengek1** [comment]
Ik weet niet of er iemand nog terug gaat komen op deze issue, maar mijn conclusie is dat het per school verschilt. Log eerst maar eens in op bijv. "ARCHIEF Pleysier College", daar krijg je een formulier met gebruikersnaam en wachtwoord tegelijk. En klik daarna bijv. op "Accent Praktijkonderwijs", daar werkt het met eerst gebruikersnaam, dan wachtwoord. Waarom dit zo is, geen flauw idee. @Fish-o

**Comment by Fish-o** [comment]
Hmm, das echt heel raar, jou PoC werkte ook niet voor mij. Ik denk dat we moeten gaan kijken of er een manier is om te checken welke nodig is, en dan beide manieren documenteren

**Comment by Underlyingglitch** [comment]
Ik ga hier straks eens even naar kijken, want ik ben hier zelf een tijd geleden mee bezig geweest en toen leek alles te werken.

**Comment by 25huizengek1** [comment]
Heeft een van jullie beiden al geprobeerd wat ik voorstelde?

**Comment by Underlyingglitch** [comment]
@25huizengek1 Had ik eerder nog niet naar gekeken. Heb zelf alleen maar te maken gehad met SSO, omdat onze school Microsoft gebruikt om in te loggen. Maar ik ga alles straks even rustig lezen en dan hoor je van me

**Comment by 25huizengek1** [comment]
Op mijn school moeten docenten ook 2fa gebruiken (verplicht), dus misschien dat dat ook voor jou geldt. Hoe kom je eigenlijk aan een docentaccount? En moet daar dan niet een andere client_id worden gebruikt?

**Comment by 25huizengek1** [comment]
In principe zou het endpoint om aan het token te komen niet eens moeten werken voor docenten, aangezien het gemaakt is voor de ouder- en leerlingapp.

**Comment by Underlyingglitch** [comment]
@25huizengek1 Nee, dat is dus meteen het probleem. Dit werkt niet voor docenten omdat deze geen mobiele app hebben.

Ik kom aan een docentaccount, omdat ik een docent ben...

**Comment by 25huizengek1** [comment]
Oh haha @Underlyingglitch , ik wist niet dat er docenten waren op Github, en al helemaal niet in een issue als deze. Als ik de ICT afdeling van mijn school zou vertellen dat we de somtoday api hebben ontcijferd zou hij me grondig gaan ondervragen...

> Na het veranderen van wat dingen aan de code (volgens mij gewoon dingen die eslint niet heel leuk vond) logt het een error, plus `Token null`, de error was `Error: Request failed with status code 302`, en als ik zo de comments lees zou dat de bedoeling moeten zijn, maar toch krijg ik geen token.

Terugdenkend aan deze comment denk ik, dat zal ik later controleren, dat een goede manier om in te loggen is dus eerst de manier van @Fish-o uit te voeren, en als daar een 302 code uit komt, kun je mijn manier proberen. Of gewoon de "bruteforce-manier" in mijn repository https://github.com/25huizengek1/somtoday-poc

De beste manier lijkt nog steeds om de gebruiker een browservenster te laten zien en net zo lang te wachten totdat er naar somtodayleerling://oauth/callback gegaan wordt. Voor docenten is het denk ik kansloos, want op de site is in ieder geval de html al door de server gesubstitueerd met de data, bijvoorbeeld huiswerk.

**Comment by Underlyingglitch** [comment]
@25huizengek1 Haha. Nou ja ik ben deels leerling en deels docent. Dat is een lang verhaal wat ik niet in een issue op GitHub ga bespreken. Als je daar meer over wilt weten, stuur me een DM op Discord (we zitten allebei in de server)

Anyways, voor docenten inloggen kan voor zover ik gezien heb niet. Omdat je eigenlijk inlogt op de app en die is er niet voor docenten. Ik heb een tijdje geleden een script geschreven om de token te verkrijgen voor mensen die dus niet technisch onderlegd zijn. Basically is het een NodeJS programma wat een aangepaste browser start en je laat inloggen bij SomToday via SSO (Microsoft) nadat je de school selecteert. Zodra de browser je wilt doorsturen naar somtodayleerling://oauth/callback dan krijg je een error in de console (immers die URL kan op een pc niet geopend worden) en die error wordt gecatcht door het programma waarna de authenticatie token opgehaald en getoond wordt.

Zie: https://github.com/Underlyingglitch/SomtodaySSOLogin

**Comment by 25huizengek1** [comment]
Ik ben eigenlijk nu wel benieuwd wat dit gaat betekenen voor de docs. Wat kunnen we het beste neerzetten? En heeft @elisaado er nog wat over te zeggen?

**Comment by elisaado** [comment]
Sinds ik op uni zit heeft @Fish-o al het onderhoud overgenomen

**Comment by 25huizengek1** [comment]
Oh excuses, dat wist ik niet. Nou, dan wend ik mij tot @Fish-o

**Comment by Underlyingglitch** [comment]
Ik vind het wel een leuk project. Heb een tijdje geleden een PR aangemaakt, maar volgens mij zit er niet veel beweging meer in het project. 

Zou graag wat actiever willen helpen

**Comment by 25huizengek1** [comment]
Enig nieuws voor wat betreft deze issue?
@Fish-o
